File: frontend/js/views/cam_reconciliation.js
// CAM reconciliation generator — commercial-lease common-area-maintenance
// true-up: pro-rata-by-sqft share of actual CAM vs estimates, via
// /calc/cam-reconciliation.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
const pct = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 }) + '%');
let LAST_DOC = null;

export async function renderCamReconciliation(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.camrec.h1.title">// CAM RECONCILIATION</span></h1>
        <p class="muted small" data-i18n="view.camrec.hint.intro">
            The annual common-area-maintenance true-up under a commercial lease. The tenant pays monthly CAM
            estimates; at year end the tenant's pro-rata share (by rentable square footage) of actual CAM is
            reconciled against what was paid, and the difference is billed or credited. An optional cap limits
            the increase over the prior year. Drafting aid, not legal/accounting advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.camrec.h2.inputs">Reconciliation inputs</h2>
            <form id="cam-form" class="inline-form">
                <label><span data-i18n="view.camrec.label.state">State</span>
                    <input type="text" name="state" value="Delaware" required></label>
                <label><span data-i18n="view.camrec.label.landlord">Landlord</span>
                    <input type="text" name="landlord_name" value=""></label>
                <label><span data-i18n="view.camrec.label.tenant">Tenant</span>
                    <input type="text" name="tenant_name" value=""></label>
                <label><span data-i18n="view.camrec.label.property">Premises</span>
                    <input type="text" name="property_label" value="Suite 100"></label>
                <label><span data-i18n="view.camrec.label.tenant_sqft">Tenant sq ft</span>
                    <input type="number" step="100" min="0" name="tenant_sqft" value="5000" required></label>
                <label><span data-i18n="view.camrec.label.building_sqft">Building sq ft</span>
                    <input type="number" step="100" min="0" name="building_sqft" value="50000" required></label>
                <label><span data-i18n="view.camrec.label.actual">Actual CAM ($)</span>
                    <input type="number" step="1000" min="0" name="actual_cam_usd" value="400000" required></label>
                <label><span data-i18n="view.camrec.label.estimate">Monthly estimate ($)</span>
                    <input type="number" step="100" min="0" name="monthly_estimate_usd" value="3000" required></label>
                <label><span data-i18n="view.camrec.label.months">Months</span>
                    <input type="number" step="1" min="1" name="months" value="12" required></label>
                <label><span data-i18n="view.camrec.label.cap">Cap on increase (%, 0 = none)</span>
                    <input type="number" step="0.1" min="0" name="cap_pct" value="0"></label>
                <label><span data-i18n="view.camrec.label.prior">Prior-year share ($)</span>
                    <input type="number" step="1000" min="0" name="prior_year_share_usd" value="0"></label>
                <label><span data-i18n="view.camrec.label.year">Reconciliation year</span>
                    <input type="text" name="year" value="2025" required></label>
                <label><span data-i18n="view.camrec.label.date">Statement date</span>
                    <input type="date" name="date" value="2026-07-01" required></label>
                <label><span data-i18n="view.camrec.label.statute">Statute citation (optional)</span>
                    <input type="text" name="statute_citation" value="" placeholder="${esc(t('view.camrec.ph.statute'))}"></label>
            </form>
        </div>
        <div id="cam-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#cam-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            landlord_name: (fd.get('landlord_name') || '').trim(),
            tenant_name: (fd.get('tenant_name') || '').trim(),
            property_label: (fd.get('property_label') || '').trim(),
            tenant_sqft: Number(fd.get('tenant_sqft')) || 0,
            building_sqft: Number(fd.get('building_sqft')) || 0,
            actual_cam_usd: Number(fd.get('actual_cam_usd')) || 0,
            monthly_estimate_usd: Number(fd.get('monthly_estimate_usd')) || 0,
            months: Number(fd.get('months')) || 0,
            cap_pct: Number(fd.get('cap_pct')) || 0,
            prior_year_share_usd: Number(fd.get('prior_year_share_usd')) || 0,
            year: (fd.get('year') || '').trim(),
            date: fd.get('date'),
            state: (fd.get('state') || '').trim(),
            statute_citation: (fd.get('statute_citation') || '').trim(),
        };
        try {
            const doc = await api.calcCamReconciliation(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.camrec.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);
    form.addEventListener('input', () => { live(); });
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase()];
    if (doc.statutory_citation) lines.push(doc.statutory_citation);
    lines.push('');
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const owes = doc.balance_usd >= 0;
    const balLabel = owes ? t('view.camrec.card.owes') : t('view.camrec.card.credit');
    const capCard = doc.cap_applied
        ? `<div class="card"><div class="label" data-i18n="view.camrec.card.uncapped">Uncapped share</div>
               <div class="value">${money(doc.tenant_share_uncapped_usd)}</div></div>`
        : '';
    const el = mount.querySelector('#cam-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card ${owes ? 'neg' : 'pos'}"><div class="label">${esc(balLabel)}</div>
                    <div class="value">${money(Math.abs(doc.balance_usd))}</div></div>
                <div class="card"><div class="label" data-i18n="view.camrec.card.share">Pro-rata share</div>
                    <div class="value">${pct(doc.pro_rata_pct)}</div></div>
                <div class="card"><div class="label" data-i18n="view.camrec.card.tenant_share">Tenant CAM share</div>
                    <div class="value">${money(doc.tenant_share_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.camrec.card.paid">Estimates paid</div>
                    <div class="value">${money(doc.estimates_paid_usd)}</div></div>
                ${capCard}
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="cam-copy" type="button" data-i18n="view.camrec.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="cam-download" type="button" data-i18n="view.camrec.btn.download">Download .txt</button>
            </div>
        </div>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#cam-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.camrec.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.camrec.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#cam-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'cam-reconciliation.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/cam_reconciliation.rs
//! CAM reconciliation — the annual common-area-maintenance true-up under a
//! commercial lease. The tenant pays monthly CAM estimates during the year; at
//! year end the landlord reconciles each tenant's pro-rata share (by rentable
//! square footage) of the actual CAM spend against what the tenant paid, and
//! bills or credits the difference. An optional cap limits the tenant's share to
//! a percentage increase over the prior year (a controllable-expense cap). No
//! existing generator computes pro-rata-by-square-foot expense reconciliation.
//! Drafting aid, not legal/accounting advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct CamInput {
    pub landlord_name: String,
    pub tenant_name: String,
    #[serde(default)]
    pub property_label: String,
    /// Tenant's rentable square footage.
    pub tenant_sqft: f64,
    /// Building's total rentable square footage.
    pub building_sqft: f64,
    /// Actual total CAM spend for the year.
    pub actual_cam_usd: f64,
    /// Monthly CAM estimate the tenant paid.
    pub monthly_estimate_usd: f64,
    /// Months covered by the reconciliation (usually 12).
    pub months: u32,
    /// Cap on the increase over the prior-year share, percent (0 = no cap).
    #[serde(default)]
    pub cap_pct: f64,
    /// Tenant's prior-year reconciled share, for the cap (0 = no cap basis).
    #[serde(default)]
    pub prior_year_share_usd: f64,
    pub year: String,
    pub date: String,
    pub state: String,
    #[serde(default)]
    pub statute_citation: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct CamReconciliation {
    pub title: String,
    /// Tenant's pro-rata share of the building, percent.
    pub pro_rata_pct: f64,
    /// Pro-rata share × actual CAM, before any cap.
    pub tenant_share_uncapped_usd: f64,
    /// Tenant's share after applying the cap.
    pub tenant_share_usd: f64,
    /// True when the cap reduced the tenant's share.
    pub cap_applied: bool,
    /// Estimates the tenant paid during the period.
    pub estimates_paid_usd: f64,
    /// Positive = tenant owes; negative = credit/refund to tenant.
    pub balance_usd: f64,
    pub statutory_citation: String,
    pub clauses: Vec<DocClause>,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

pub fn generate(i: &CamInput) -> CamReconciliation {
    let share = if i.building_sqft > 0.0 {
        i.tenant_sqft / i.building_sqft
    } else {
        0.0
    };
    let uncapped = cents(i.actual_cam_usd * share);

    // Controllable-expense cap: tenant share limited to prior-year share grown
    // by at most cap_pct.
    let (tenant_share, cap_applied) = if i.cap_pct > 0.0 && i.prior_year_share_usd > 0.0 {
        let max_share = cents(i.prior_year_share_usd * (1.0 + i.cap_pct / 100.0));
        if uncapped > max_share {
            (max_share, true)
        } else {
            (uncapped, false)
        }
    } else {
        (uncapped, false)
    };

    let paid = cents(i.monthly_estimate_usd * i.months as f64);
    let balance = cents(tenant_share - paid);

    let property = if i.property_label.trim().is_empty() {
        "the Premises".to_string()
    } else {
        i.property_label.trim().to_string()
    };

    let owes = balance >= 0.0;
    let settle = if owes {
        format!("the Tenant owes the Landlord {}", money(balance.abs()))
    } else {
        format!("the Landlord shall credit the Tenant {}", money(balance.abs()))
    };

    let cap_line = if cap_applied {
        format!(
            " The uncapped share of {} was reduced to {} by the {:.1}% cap over the prior-year share of {}.",
            money(uncapped),
            money(tenant_share),
            i.cap_pct,
            money(i.prior_year_share_usd)
        )
    } else {
        String::new()
    };

    let calc_body = format!(
        "The Tenant's pro-rata share is {:.0} ÷ {:.0} rentable square feet = {:.2}%. Applied to actual CAM of {} for {}, the Tenant's share is {}.{} The Tenant paid {} in monthly estimates ({} × {} months), so {}.",
        i.tenant_sqft,
        i.building_sqft,
        cents(share * 100.0),
        money(i.actual_cam_usd),
        i.year,
        money(tenant_share),
        cap_line,
        money(paid),
        money(i.monthly_estimate_usd),
        i.months,
        settle
    );

    let citation = i.statute_citation.trim();
    let governing = if citation.is_empty() {
        format!("This reconciliation is governed by the lease and the laws of the State of {}.", i.state)
    } else {
        format!("This reconciliation is governed by the lease and the laws of the State of {} ({}).", i.state, citation)
    };

    let clauses = vec![
        DocClause {
            heading: "Parties".into(),
            body: format!(
                "Landlord: {}\nTenant: {}\nPremises: {}\nReconciliation year: {}\nStatement date: {}",
                i.landlord_name, i.tenant_name, property, i.year, i.date
            ),
        },
        DocClause {
            heading: "1. Pro-Rata Share".into(),
            body: format!(
                "The Tenant's pro-rata share of common-area maintenance is {:.2}%, being the Tenant's {:.0} rentable square feet divided by the building's {:.0} rentable square feet.",
                cents(share * 100.0),
                i.tenant_sqft,
                i.building_sqft
            ),
        },
        DocClause { heading: "2. Reconciliation".into(), body: calc_body },
        DocClause {
            heading: "3. Payment".into(),
            body: if owes {
                format!("The Tenant shall pay the reconciled balance of {} within 30 days of this statement.", money(balance.abs()))
            } else {
                format!("The Landlord shall apply the overpayment credit of {} to the next CAM estimate or refund it within 30 days.", money(balance.abs()))
            },
        },
        DocClause {
            heading: "4. Audit Right".into(),
            body: "The Tenant may, within 90 days of this statement and on reasonable notice, audit the Landlord's CAM records for the reconciliation year. Overcharges revealed by the audit shall be refunded.".into(),
        },
        DocClause { heading: "5. Governing Law".into(), body: governing },
        DocClause {
            heading: "Signatures".into(),
            body: format!(
                "Landlord: ____________________  Date: __________\n{}\n\nTenant: ____________________  Date: __________\n{}",
                i.landlord_name, i.tenant_name
            ),
        },
    ];

    CamReconciliation {
        title: "CAM Reconciliation Statement".into(),
        pro_rata_pct: cents(share * 100.0),
        tenant_share_uncapped_usd: uncapped,
        tenant_share_usd: cents(tenant_share),
        cap_applied,
        estimates_paid_usd: paid,
        balance_usd: balance,
        statutory_citation: citation.to_string(),
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> CamInput {
        CamInput {
            landlord_name: "Plaza Owners LP".into(),
            tenant_name: "Cafe Tenant LLC".into(),
            property_label: "Suite 100".into(),
            tenant_sqft: 5_000.0,
            building_sqft: 50_000.0,
            actual_cam_usd: 400_000.0,
            monthly_estimate_usd: 3_000.0,
            months: 12,
            cap_pct: 0.0,
            prior_year_share_usd: 0.0,
            year: "2025".into(),
            date: "2026-07-01".into(),
            state: "Delaware".into(),
            statute_citation: String::new(),
        }
    }

    #[test]
    fn tenant_owes_underpayment() {
        let d = generate(&base());
        assert!(close(d.pro_rata_pct, 10.0));
        assert!(close(d.tenant_share_usd, 40_000.0));
        assert!(close(d.estimates_paid_usd, 36_000.0));
        assert!(close(d.balance_usd, 4_000.0));
        assert!(!d.cap_applied);
    }

    #[test]
    fn overpayment_is_credit() {
        let d = generate(&CamInput { monthly_estimate_usd: 3_500.0, ..base() });
        // paid 42,000 vs share 40,000 → -2,000 credit.
        assert!(close(d.balance_usd, -2_000.0));
        assert!(d.clauses.iter().any(|c| c.body.contains("credit")));
    }

    #[test]
    fn cap_limits_share_over_prior_year() {
        let d = generate(&CamInput { cap_pct: 5.0, prior_year_share_usd: 35_000.0, ..base() });
        // max share = 35,000 × 1.05 = 36,750 < 40,000 uncapped.
        assert!(close(d.tenant_share_uncapped_usd, 40_000.0));
        assert!(close(d.tenant_share_usd, 36_750.0));
        assert!(d.cap_applied);
        assert!(close(d.balance_usd, 750.0));
    }

    #[test]
    fn cap_not_applied_when_share_below_max() {
        let d = generate(&CamInput { cap_pct: 25.0, prior_year_share_usd: 35_000.0, ..base() });
        // max = 43,750 > 40,000 → no cap.
        assert!(!d.cap_applied);
        assert!(close(d.tenant_share_usd, 40_000.0));
    }

    #[test]
    fn zero_building_sqft_no_divide_by_zero() {
        let d = generate(&CamInput { building_sqft: 0.0, ..base() });
        assert!(close(d.pro_rata_pct, 0.0));
        assert!(close(d.tenant_share_usd, 0.0));
    }

    #[test]
    fn audit_right_clause_present() {
        assert!(generate(&base()).clauses.iter().any(|c| c.heading.contains("Audit")));
    }

    #[test]
    fn statute_citation_echoed() {
        let d = generate(&CamInput { statute_citation: "Cal. Civ. Code § 1950.8".into(), ..base() });
        assert_eq!(d.statutory_citation, "Cal. Civ. Code § 1950.8");
        assert!(d.clauses.iter().any(|c| c.body.contains("Cal. Civ. Code § 1950.8")));
    }
}
